runtime/common/src/crowdloan.rs

```diff
@@ -465,52 +448,79 @@ decl_module! {
 
 			Self::contribution_kill(fund.trie_index, &who);
 			fund.raised = fund.raised.saturating_sub(balance);
-			if !fund.retiring {
-				fund.retiring = true;
-				fund.end = now;
-			}
 
 			Funds::<T>::insert(index, &fund);
 
 			Self::deposit_event(RawEvent::Withdrew(who, index, balance));
 		}
 
-		/// Remove a fund after the retirement period has ended and all funds have been returned.
+		/// Automatically refund contributors of an ended crowdloan.
+		/// Due to weight restrictions, this function may need to be called multiple
+		/// times to fully refund all users. We will refund `RemoveKeysLimit` users at a time.
 		///
-		/// This places any deposits that were not withdrawn into the treasury.
-		#[weight = T::WeightInfo::dissolve(T::RemoveKeysLimit::get())]
-		pub fn dissolve(origin, #[compact] index: ParaId) -> DispatchResultWithPostInfo {
-			let who = ensure_signed(origin)?;
+		/// Origin must be signed, but can come from anyone.
+		#[weight = T::WeightInfo::refund(T::RemoveKeysLimit::get())]
+		pub fn refund(origin, #[compact] index: ParaId) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
 
-			let fund = Self::funds(index).ok_or(Error::<T>::InvalidParaId)?;
+			let mut fund = Self::funds(index).ok_or(Error::<T>::InvalidParaId)?;
 			let now = frame_system::Pallet::<T>::block_number();
-			let dissolution = fund.end.saturating_add(T::RetirementPeriod::get());
-
-			let can_dissolve = (fund.retiring && now >= dissolution) ||
-				(fund.raised.is_zero() && who == fund.depositor);
-			ensure!(can_dissolve, Error::<T>::NotReadyToDissolve);
+			let fund_account = Self::fund_account_id(index);
+			Self::ensure_crowdloan_ended(now, &fund_account, &fund)?;
 
+			let mut refund_count = 0u32;
 			// Try killing the crowdloan child trie
-			match Self::crowdloan_kill(fund.trie_index) {
-				child::KillChildStorageResult::AllRemoved(num_removed) => {
-					CurrencyOf::<T>::unreserve(&fund.depositor, fund.deposit);
+			let contributions = Self::contribution_iterator(fund.trie_index);
+			// Assume everyone will be refunded.
+			let mut all_refunded = true;
+			for (who, (balance, _)) in contributions {
+				if refund_count >= T::RemoveKeysLimit::get() {
+					// Not everyone was able to be refunded this time around.
+					all_refunded = false;
+					break;
+				}
+				CurrencyOf::<T>::transfer(&fund_account, &who, balance, AllowDeath)?;
+				Self::contribution_kill(fund.trie_index, &who);
+				fund.raised = fund.raised.saturating_sub(balance);
+				refund_count += 1;
+			}
```

Comment:
```suggestion
			let contributions = Self::contribution_iterator(fund.trie_index);
			let refunded = contributions.into_iter().take(T::RemoveKeysLimit::get()).fold(Zero::zero(), |refunded, (who, (balance, _))| {
				CurrencyOf::<T>::transfer(&fund_account, &who, balance, AllowDeath)?;
				Self::contribution_kill(fund.trie_index, &who);
				fund.raised = fund.raised.saturating_sub(balance);
				refunded += 1 
			});
			
			if refunded == T::RemoveKeysLimit::get() {
			  // everyone was refunded
			} else {
			  // some left 
			}
```

I find this simpler but just a suggestion